content/start/guess_the_correlation/index.Rmarkdown

```diff
@@ -0,0 +1,632 @@
+      0.0000  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999
+      0.0000  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999  0.9999
+    ... [the output was truncated (use n=-1 to disable)]
+    [ CPUFloatType{1,130,130} ]
+
+    $y
+    torch_tensor
+    -0.45781
+    [ CPUFloatType{} ]
+
+    $id
+    [1] "arjskzyc"
+
+It's a list of three items, the last of which we're not interested in for our purposes.
+
+The second, a scalar tensor, is the correlation value, the thing we want the network to learn. The first, `x`, is the scatterplot: a tensor representing an image of dimensionality 130\*130. But wait -- what is that `1` in the shape output?
+
+    [ CPUFloatType{1,130,130} ]
+
+This really is a three-dimensional tensor! The first dimension holds different *channels* -- or the single channel, if the image has but one. In fact, the reason `x` came in this format is that we requested it, here:
+
+```{r, eval=FALSE}
+add_channel_dim <- function(img) img$unsqueeze(1)
+
+train_ds <- guess_the_correlation_dataset(
+    # ...
+    transform = function(img) crop_axes(img) %>% add_channel_dim(),
+    # ...
+  )
+```
+
+`add_channel_dim()` was passed in as a custom transformation, to be applied to every item of the dataset. It calls one of `torch`'s many tensor operations, `unsqueeze()`, that adds a singleton dimension at a requested position.
+
+How about the second custom transformation?
+
+```{r, eval=FALSE}
+crop_axes <- function(img) transform_crop(img, top = 0, left = 21, height = 131, width = 130)
+```
+
+Here, we crop the image, cutting off the axes and labels on the left and bottom. These image regions don't contribute any distinctive information, and having the images be smaller saves memory.
+
+# Work with batches
+
+Now, we've done so much work already, but you haven't actually *seen* any of the scatterplots yet! The reason we've been waiting until now is that we want to show a bunch of them at a time, and for that, we need to know how to handle *batches* of data.
+
+So let's create a `DataLoader` object from the training set. We'll soon use it to train the model, but right now, we'll just plot the first batch.
+
+A `DataLoader` needs to know where to get the data -- namely, from the `Dataset` it gets passed --, as well as how many items should go in a batch. Optionally, it can return data in random order (`shuffle = TRUE`).
+
+```{r}
+train_dl <- dataloader(train_ds, batch_size = 64, shuffle = TRUE)
+```
+
+Like a `Dataset`, we can query a `DataLoader` for its length. For the `Dataset`, this meant number of items; for a `DataLoader` , it means number of batches:
+
+```{r}
+length(train_dl)
+```
+
+    [1] 157
+
+To access the first batch, we create an iterator from the `DataLoader` and ask it for the first batch. Even if it weren't for plotting, you might do this just to check that the dimensions look ok:
+
+```{r}
+iter <- train_dl$.iter()
```

Comment:
can we use `dataloader_make_iter` and `dataloader_next` here?

Reply:
oh yeah, for sure, thank you!